tech-specs/src/specs/ledger.md

```diff
@@ -0,0 +1,190 @@
+1. Inject a [stack height](#stack-height-limiter) limiter into the `code`.
+1. Compile the validity predicate with single-pass compiler. The compilation computational complexity MUST be linear in proportion to its bytes size.
+1. Initialize the WASM linear memory with descriptor having the initial memory size equal to [`VP_MEMORY_INIT_PAGES`](#wasm-constants) and maximum memory size to [`VP_MEMORY_MAX_PAGES`](#wasm-constants).
+1. Instantiate the WASM module with imported [validity predicate host environment functions](#validity-predicate-host-environment-functions) and the instantiated WASM memory.
+1. Write the address of the validity predicate’s owner, the transaction `data`, the modified storage keys encoded with Borsh, and all the triggered validity predicates owners' addresses encoded with Borsh into the memory exported from the WASM module instance.
+1. Attempt to call the module's entrypoint function. The entrypoint MUST have signature:
+   ```wat
+   func (param i64 i64 i64 i64 i64 i64 i64 i64) (result i64))
+   ```
+   - The first argument is the offset to the owner’s address written into the memory, the second argument is its bytes length
+   - The third is the offset of the transaction’s `data` and fourth is it’s bytes length
+   - The fifth is the offset of the modified storage keys and sixth is its bytes length
+   - The seventh is the offset of the triggered validity predicates owners' addresses and eighth is its bytes length
+
+#### Gas
+
+##### Gas constants
+
+The gas constants are currently chosen arbitrarily and are subject to change following gas accounting estimations.
+
+| Name                   | Value |
+|------------------------|-------|
+| `COMPILE_GAS_PER_BYTE` |     1 |
+| `BASE_TRANSACTION_FEE` |     2 |
+| `PARALLEL_GAS_DIVIDER` |    10 |
+| `MIN_STORAGE_GAS`      |     1 |
+
+- TODO describe gas accounting, wasm gas counter, limits, what happens if we go over limits and how gas relates to fees
+
+#### WebAssembly (WASM)
+
+##### WASM constants
+
+| Name                                 | Unit              | Value |
+|--------------------------------------|-------------------|-------|
+| `PAGE` (as defined in the WASM spec) | kiB               |    64 |
+| `TX_MEMORY_INIT_PAGES`               | number of `PAGE`s |   100 |
+| `TX_MEMORY_MAX_PAGES`                | number of `PAGE`s |   200 |
+| `VP_MEMORY_INIT_PAGES`               | number of `PAGE`s |   100 |
+| `VP_MEMORY_MAX_PAGES`                | number of `PAGE`s |   200 |
+| `WASM_STACK_LIMIT`                   | stack depth       | 65535 |
+
+
+The WASM instantiation, the types, instructions, validation and execution of WASM modules MUST conform to the [WebAssembly specification](https://webassembly.github.io/spec/core/intro/index.html).
+
+##### WASM validation
+
+The WebAssembly code is REQUIRED to only use deterministic instructions. Furthermore, it MUST NOT use features from any of the following WebAssembly proposals:
```

Comment:
and presumably we validate this?

Reply:
yup, the tx and VP sections link to this with `1. [Validate the WASM code](#wasm-validation) from the `code` field of the transaction.` and `1. [Validate the WASM code](#wasm-validation) of the validity predicate.`, resp.